### ISSUE TITLE
linspace->float-substitution fails with complex numbers

### ISSUE THREAD

**Issue by afranson** [numcl/numcl#62]
When running
```
(numcl:linspace #C(1d0 1d0) #C(1d0 4d0) :num 4)
```
, I get the following error from sly,
```
Pattern: ((T T) (NIL NIL)
          ((TYPE-R:FLOAT-TYPE)
           NUMCL.IMPL::*NUMCL-DEFAULT-FLOAT-FORMAT*)
          ((TYPE-R:LONG-FLOAT-TYPE) 'LONG-FLOAT)
          ((TYPE-R:DOUBLE-FLOAT-TYPE) 'DOUBLE-FLOAT)
          ((TYPE-R:SINGLE-FLOAT-TYPE) 'SINGLE-FLOAT)
          ((TYPE-R:SHORT-FLOAT-TYPE) 'SHORT-FLOAT)
          ((TYPE-R:INTEGER-SUBTYPE) NUMCL.IMPL::INT-RESULT)
          ((TYPE-R:RATIO-TYPE)
           NUMCL.IMPL::*NUMCL-DEFAULT-FLOAT-FORMAT*)
          ((TYPE-R:REAL-SUBTYPE)
           (ERROR "this should not happen"))
          ((TYPE-R:COMPLEX-TYPE NUMCL.IMPL::ELEMENT-TYPE)
           `(COMPLEX
             ,(NUMCL.IMPL::FLOAT-SUBSTITUTION
               NUMCL.IMPL::ELEMENT-TYPE :INT-RESULT
               NUMCL.IMPL::*NUMCL-DEFAULT-FLOAT-FORMAT*)))
          ((TYPE-R:OR-TYPE NUMCL.IMPL::TYPES)
           `(OR
             ,@(REMOVE-DUPLICATES
                (REMOVE NIL
                        #1=(MAPCAR #'NUMCL.IMPL::REC
                                   NUMCL.IMPL::TYPES)))))
          ((TYPE-R:AND-TYPE NUMCL.IMPL::TYPES)
           (TRIVIA.LEVEL2:MATCH #1#
             ((LIST* FIRST REST)
              (IF (EVERY (ALEXANDRIA:CURRY #'EQ FIRST) REST)
                  FIRST
                  NIL))))) 
 Values: ((TYPE-R:COMPLEX-TYPE DOUBLE-FLOAT))
   [Condition of type TRIVIA.LEVEL2:MATCH-ERROR]
```
. It fails with double and single float inputs.

**Comment by afranson** [comment]
Seems :num = 4 was a bit of a special case. For most other values (for example 5 or the default, 50), the error and backtrace is
```
this should not happen
   [Condition of type SIMPLE-ERROR]

Backtrace:
 0: ((FLET "FN541" :IN NUMCL.IMPL::FLOAT-SUBSTITUTION))
 1: ((LABELS NUMCL.IMPL::REC :IN NUMCL.IMPL::FLOAT-SUBSTITUTION) (RATIONAL 0 3/4))
 2: ((LABELS NUMCL.IMPL::REC :IN NUMCL.IMPL::FLOAT-SUBSTITUTION) (COMPLEX (RATIONAL 0 3/4)))
 3: (NUMCL.IMPL::FLOAT-CONTAGION (COMPLEX (RATIONAL 0 3/4)) (TYPE-R:COMPLEX-TYPE SINGLE-FLOAT) :INT-INT-RESULT INTEGER)
 4: (NUMCL.EXPORTED:LINSPACE #C(1 1) #C(1 4) :NUM 5 :ENDPOINT T :TYPE NIL)
```

**Comment by afranson** [comment]
Even once it passes the type coercion, linspace will fail again in the loop macro because the step does not satisfy the (real 0) type requirement of the loop. You might be able to override that behavior, but it may be simpler to just replace it with a brute force do.
```
(declaim (inline linspace))
(defun linspace (start stop &key (num 50) (endpoint t) type)
  (declare (fixnum num))
  (declare (number start stop))
  (let* ((step (/ (- stop start) (1- num)))
	 (type (or type (%arange-infer-type start stop step)))
         (start (coerce start type))
         (stop  (coerce stop  type))
         (step  (coerce step  type))
         (lst (do ((n 1 (1+ n))
		   (x (list start) (push (+ step (car x)) x)))
		  ((>= n (1- num)) (reverse x))))
         (lst (if endpoint (append lst (list stop)) lst))
	 (arr (asarray lst :type type)))
    (values arr step)))
```

**Comment by afranson** [comment]
I think I solved it. I will be making a pull request shortly. With my modifications, linspace and asarray now work with complex numbers - except for when the complex numbers contain rational numbers.

**Comment by afranson** [comment]
Solved with pull #63.
